tactics/tactics.ml

```diff
@@ -4987,15 +4987,15 @@ let anon_id = Id.of_string "anonymous"
 let name_op_to_name name_op object_kind suffix =
   let open Proof_global in
   let default_gk = (Global, false, object_kind) in
+  let name, gk = match Proof_global.V82.get_current_initial_conclusions () with
+  | (id, (_, gk)) -> Some id, gk
+  | exception NoCurrentProof -> None, default_gk
+  in
   match name_op with
-    | Some s ->
-      (try let _, gk, _ = Pfedit.current_proof_statement () in s, gk
-       with NoCurrentProof -> s, default_gk)
-    | None   ->
-      let name, gk =
-	try let name, gk, _ = Pfedit.current_proof_statement () in name, gk
-	with NoCurrentProof -> anon_id, default_gk in
-      add_suffix name suffix, gk
+  | Some s -> s, gk
+  | None ->
+    let name = Option.default anon_id name in
+    add_suffix name suffix, gk
```

Comment:
Uh, while you are at it, do you think it'd be possible to remove the dependency on global state here?

Reply:
Not sure how to do that easily, albeit desirable. There is a hierarchical dependency breakage here, abstract should not care about the name of the proof being generated.

Reply:
Yeah that's one of the very few blocking spots that the proof global removal patch has.

Reply:
A easy fix would be to generate a lemma-independent name using some kind of nonce. I believe that if you rely on the names of generated subproofs, you deserve to die. Also, a more radical solution would be to implement those godamn definition-local values that would properly solve the abstract side-effects hell.

Reply:
> you deserve to die.

Or to thread an argument around in all your functions. I dunno which is worse.

Reply:
I can try something with a uniquely generated deterministic name, like e.g. taking the hash of the goal type and taint it with the goal evar number or something.

Reply:
I've added your comments to my working patch list, I suggest you wait for a first candidate version of the Proof_global removal patch and then we see what we can do there, maybe there is a more convenient way.

Reply:
> I can try something with a uniquely generated deterministic name, like e.g. taking the hash of the goal type and taint it with the goal evar number or something.

Uh, I'd rather not that.  I can just imagine chasing down bugs of the form "abstract only succeeds if I create and then clear an evar just before creating the goal in which abstract is run" or something insane like that.

Reply:
The irony is that the above approach would be less broken than the current one [which works only because the parallel proof checking engine is not very capable as of today]